mava/systems/ppo/sebulba/rec_ippo.py

```diff
@@ -0,0 +1,897 @@
+    rollout_queue: Pipeline,
+    params_source: ParamsSource,
+    apply_fns: Tuple[RecActorApply, RecCriticApply],
+    actor_device: int,
+    seeds: List[int],
+    thread_lifetime: ThreadLifetime,
+) -> None:
+    """Runs rollouts to collect trajectories from the environment.
+
+    Args:
+        key (chex.PRNGKey): The PRNGkey.
+        config (DictConfig): Configuration settings for the environment and rollout.
+        rollout_queue (Pipeline): Queue for sending collected rollouts to the learner.
+        params_source (ParamsSource): Source for fetching the latest network parameters
+        from the learner.
+        apply_fns (Tuple): Functions for running the actor and critic networks.
+        actor_device (Device): Actor device to use for rollout.
+        seeds (List[int]): Seeds for initializing the environment.
+        thread_lifetime (ThreadLifetime): Manages the thread's lifecycle.
+    """
+    name = threading.current_thread().name
+    print(f"{Fore.BLUE}{Style.BRIGHT}Thread {name} started{Style.RESET_ALL}")
+    actor_apply_fn, critic_apply_fn = apply_fns
+    num_agents, num_envs = config.system.num_agents, config.arch.num_envs
+    move_to_device = lambda x: jax.device_put(x, device=actor_device)
+
+    @jax.jit
+    def act_fn(
+        params: Params,
+        observation: Observation,
+        dones: chex.Array,
+        hstates: HiddenStates,
+        key: chex.PRNGKey,
+    ) -> Tuple:
+        """Get action and value."""
+        # actor_policy = actor_apply_fn(params.actor_params, observation)
+
+        batched_observation = tree.map(lambda x: x[jnp.newaxis, :], observation)
+        ac_in = (
+            batched_observation,
+            dones[jnp.newaxis, :],
+        )
+        policy_hidden_state, actor_policy = actor_apply_fn(
+            params.actor_params, hstates.policy_hidden_state, ac_in
+        )
+        critic_hidden_state, value = critic_apply_fn(
+            params.critic_params, hstates.critic_hidden_state, ac_in
+        )
+
+        action = actor_policy.sample(seed=key)
+        log_prob = actor_policy.log_prob(action)
+        # It may be faster to calculate the values in the learner as
+        # then we won't need to pass critic params to actors.
+        # value = critic_apply_fn(params.critic_params, observation).squeeze()
+
+        hstates = HiddenStates(policy_hidden_state, critic_hidden_state)
+        return action, log_prob, value, hstates
+
+    timestep = env.reset(seed=seeds)
+    dones = np.repeat(timestep.last(), num_agents).reshape(num_envs, -1)
+
+    # Initialise hidden states.
+    init_policy_hstate = ScannedRNN.initialize_carry(
+        (config.arch.num_envs, num_agents), config.network.hidden_state_dim
+    )
+    init_critic_hstate = ScannedRNN.initialize_carry(
+        (config.arch.num_envs, num_agents), config.network.hidden_state_dim
+    )
+    hstates = HiddenStates(init_policy_hstate, init_critic_hstate)
+    hstates_tpu = tree.map(move_to_device, hstates)
```

Comment:
`device_put` already tree maps
```suggestion
    hstates_tpu = move_to_device(hstates)
```